File: Cosmos-IOS/Cosmostation/Cosmostation/Utils/StarnameAssets.swift
//
//  StarnameAssets.swift
//  Cosmostation
//
//  Created by 정용주 on 2021/10/12.
//  Copyright © 2021 wannabit. All rights reserved.
//

import Foundation

public func getStarNameChainImgUrl(_ uri: String?) -> URL {
    if let asset = getStarnameAssets().filter { $0.uri == uri }.first {
        return URL(string: asset.chainImg ?? "")!
    }
    return URL(string: CHAIN_IMG_URL + "unknown.png")!
}

public func getStarNameChainName(_ uri: String?) -> String {
    if let asset = getStarnameAssets().filter { $0.uri == uri }.first {
        return asset.name
    }
    return "Unknown"
}
    

public func getStarnameAssets() -> Array<StarnameAsset> {
    var result = Array<StarnameAsset>()
    result.append(.ada)
    result.append(.akt)
    result.append(.algo)
    result.append(.atom)
    result.append(.avax)
    result.append(.band)
    result.append(.bat)
    result.append(.bch)
    result.append(.bnb)
    result.append(.btc)
    result.append(.comp)
    result.append(.cro)
    result.append(.ctk)
    result.append(.dai)
    result.append(.dash)
    result.append(.doge)
    result.append(.dot)
    result.append(.dsm)
    result.append(.dvpn)
    result.append(.emy)
    result.append(.eos)
    result.append(.eth)
    result.append(.ion)
    result.append(.iov)
    result.append(.iris)
    result.append(.kava)
    result.append(.ksm)
    result.append(.link)
    result.append(.lsk)
    result.append(.ltc)
    result.append(.neo)
    result.append(.okb)
    result.append(.osmo)
    result.append(.regen)
    result.append(.ren)
    result.append(.req)
    result.append(.rowan)
    result.append(.scrt)
    result.append(.shib)
    result.append(.sol)
    result.append(.svt)
    result.append(.ter)
    result.append(.trx)
    result.append(.usdc)
    result.append(.usdt)
    result.append(.xlm)
    result.append(.xmr)
    result.append(.xprt)
    result.append(.xtz)
    result.append(.zec)
    result.append(.wbtc)
    return result
}

public struct StarnameAsset {
    public let name: String
    public let uri: String
    public let symbol: String
    public let chainImg: String?
    public let chainType: ChainType?
    
    public init(_ name: String, _ uri: String, _ symbol: String, _ chainImg: String?, _ chainType: ChainType?) {
        self.name = name
        self.uri = uri
        self.symbol = symbol
        self.chainType = chainType
        
        if (chainImg == nil) {
            self.chainImg = CHAIN_IMG_URL + "unknown.png"
        } else {
            self.chainImg = CHAIN_IMG_URL + chainImg!
        }
    }
}


public extension StarnameAsset {
    static let ada = StarnameAsset("Cardano", "asset:ada", "ADA", "ada.png", nil)
    static let akt = StarnameAsset("Akash", "asset:akt", "AKT", "akash.png", ChainType.AKASH_MAIN)
    static let algo = StarnameAsset("Algorand", "asset:algo", "ALGO", "algo.png", nil)
    static let atom = StarnameAsset("Cosmos", "asset:atom", "ATOM", "cosmos.png", ChainType.COSMOS_MAIN)
    static let avax = StarnameAsset("Avalanche", "asset:avax", "AVAX", "avax.png", nil)
    static let band = StarnameAsset("Band Protocol", "asset:band", "BAND", "band.png", ChainType.BAND_MAIN)
    static let bat = StarnameAsset("Basic Attention Token", "asset:bat", "BAT", "bat.png", nil)
    static let bch = StarnameAsset("Bitcoin Cash", "asset:bch", "BCH", "bitcoincash.png", nil)
    static let bnb = StarnameAsset("BNB coin", "asset:bnb", "BNB", "binance.png", ChainType.BINANCE_MAIN)
    static let btc = StarnameAsset("Bitcoin", "asset:btc", "BTC", "bitcoin.png", nil)
    static let comp = StarnameAsset("Compound", "asset:comp", "COMP", "comp.png", nil)
    static let cro = StarnameAsset("Crypto.org", "asset:cro", "CRO", "crypto-org.png", ChainType.CRYPTO_MAIN)
    static let ctk = StarnameAsset("CertiK", "asset:ctk", "CTK", "certik.png", ChainType.CERTIK_MAIN)
    static let dai = StarnameAsset("Dai Stablecoin", "asset:dai", "DAI", "dai.png", nil)
    static let dash = StarnameAsset("Dash", "asset:dash", "DASH", "dash.png", nil)
    static let doge = StarnameAsset("Dogecoin", "asset:doge", "DOGE", "doge.png", nil)
    static let dot = StarnameAsset("Polkadot", "asset:dot", "DOT", "dot.png", nil)
    static let dsm = StarnameAsset("Desmos", "asset:dsm", "DSM", "dsm.png", nil)
    static let dvpn = StarnameAsset("Sentinel", "asset:dvpn", "DVPN", "sentinel.png", ChainType.SENTINEL_MAIN)
    static let eos = StarnameAsset("EOS", "asset:eos", "eos", "eos.png", nil)
    static let eth = StarnameAsset("Ethereum", "asset:eth", "ETH", "ethereum.png", nil)
    static let ion = StarnameAsset("ION", "asset:ion", "ION", "osmosis.png", ChainType.OSMOSIS_MAIN)
    static let iov = StarnameAsset("Starname (IOV)", "asset:iov", "IOV", "starname.png", ChainType.IOV_MAIN)
    static let iris = StarnameAsset("IRISnet", "asset:iris", "IRIS", "iris.png", ChainType.IRIS_MAIN)
    static let kava = StarnameAsset("Kava", "asset:kava", "KAVA", "kava.png", ChainType.KAVA_MAIN)
    static let ksm = StarnameAsset("Kusama", "asset:ksm", "KSM", "ksm.png", nil)
    static let link = StarnameAsset("Chainlink", "asset:link", "LINK", "link.png", nil)
    static let ltc = StarnameAsset("Litecoin", "asset:ltc", "LTC", "litecoin.png", nil)
    static let neo = StarnameAsset("Neo", "asset:neo", "NEO", "neo.png", nil)
    static let okb = StarnameAsset("OKB", "asset:okb", "OKB", "okex.png", ChainType.OKEX_MAIN)
    static let osmo = StarnameAsset("Osmosis", "asset:osmo", "OSMO", "osmosis.png", ChainType.OSMOSIS_MAIN)
    static let regen = StarnameAsset("Regen Network", "asset:regen", "REGEN", "regen.png", ChainType.REGEN_MAIN)
    static let ren = StarnameAsset("Ren", "asset:ren", "REN", "ren.png", nil)
    static let req = StarnameAsset("Request", "asset:req", "REQ", "req.png", nil)
    static let rowan = StarnameAsset("Sifchain", "asset:rowan", "ROWAN", "sifchain.png", ChainType.SIF_MAIN)
    static let scrt = StarnameAsset("Secret", "asset:scrt", "SCRT", "secret.png", ChainType.SECRET_MAIN)
    static let shib = StarnameAsset("Shiba Inu", "asset:shib", "SHIB", "shib.png", nil)
    static let svt = StarnameAsset("Savitar Token", "asset:svt", "SVT", "svt.png", nil)
    static let sol = StarnameAsset("Solana", "asset:sol", "SOL", "sol.png", nil)
    static let trx = StarnameAsset("TRON", "asset:trx", "TRX", "trx.png", nil)
    static let usdc = StarnameAsset("USD Coin", "asset:usdc", "USDC", "usdc.png", nil)
    static let usdt = StarnameAsset("Tether", "asset:usdt", "USDT", "usdt.png", nil)
    static let xlm = StarnameAsset("Stellar", "asset:xlm", "XLM", "xlm.png", nil)
    static let xmr = StarnameAsset("Monero", "asset:xmr", "XMR", "xmr.png", nil)
    static let xprt = StarnameAsset("Persistence", "asset:xprt", "XPRT", "persistence.png", ChainType.PERSIS_MAIN)
    static let xtz = StarnameAsset("Tezos", "asset:xtz", "XTZ", "tezos.png", nil)
    static let zec = StarnameAsset("Zcash", "asset:zec", "ZEC", "zec.png", nil)
    static let wbtc = StarnameAsset("Wrapped Bitcoin", "asset:wbtc", "WBTC", "wbtc.png", nil)
    
    static let ter = StarnameAsset("Terra", "asset:luna", "LUNA", "terra.png", nil)
    static let emy = StarnameAsset("E-Money", "asset:ngm", "NGM", "emoney.png", nil)
    static let lsk = StarnameAsset("LISK", "asset:lsk", "LSK", "lsk.png", nil)
}

File: Cosmos-Android/app/src/main/java/wannabit/io/cosmostaion/utils/StarnameAssets.java
package wannabit.io.cosmostaion.utils;

import java.util.ArrayList;

import wannabit.io.cosmostaion.base.BaseChain;

import static wannabit.io.cosmostaion.base.BaseConstant.CHAIN_IMG_URL;

public class StarnameAssets {
    public String name;
    public String url;
    public String symbol;
    public String chainImg;
    public BaseChain baseChain;

    public StarnameAssets(String name, String url, String symbol, String chainImg, BaseChain baseChain) {
        this.name = name;
        this.url = url;
        this.symbol = symbol;
        this.baseChain = baseChain;

        if (chainImg == null) {
            this.chainImg = CHAIN_IMG_URL + "unknown.png";
        } else {
            this.chainImg = CHAIN_IMG_URL + chainImg;
        }
    }

    public static ArrayList<StarnameAssets> getStarnameAssets() {
        ArrayList<StarnameAssets> result = new ArrayList<>();
        result.add(ada);
        result.add(akt);
        result.add(algo);
        result.add(atom);
        result.add(avax);
        result.add(band);
        result.add(bat);
        result.add(bch);
        result.add(bnb);
        result.add(btc);
        result.add(comp);
        result.add(cro);
        result.add(dai);
        result.add(dash);
        result.add(doge);
        result.add(dsm);
        result.add(dvpn);
        result.add(eos);
        result.add(eth);
        result.add(ion);
        result.add(iov);
        result.add(iris);
        result.add(kava);
        result.add(ksm);
        result.add(link);
        result.add(ltc);
        result.add(neo);
        result.add(okb);
        result.add(osmo);
        result.add(regen);
        result.add(ren);
        result.add(req);
        result.add(rowan);
        result.add(scrt);
        result.add(shib);
        result.add(sol);
        result.add(svt);
        result.add(trx);
        result.add(usdc);
        result.add(usdt);
        result.add(wbtc);
        result.add(xlm);
        result.add(xmr);
        result.add(xprt);
        result.add(xtz);
        result.add(zec);
        return result;
    }

    public static StarnameAssets ada    = new StarnameAssets("Cardano", "asset:ada", "ADA", null, null);
    public static StarnameAssets akt    = new StarnameAssets("Akash", "asset:akt", "AKT", "akash.png", BaseChain.AKASH_MAIN);
    public static StarnameAssets algo   = new StarnameAssets("Algorand", "asset:algo", "ALGO", null, null);
    public static StarnameAssets atom   = new StarnameAssets("Cosmos", "asset:atom", "ATOM", "cosmos.png", BaseChain.COSMOS_MAIN);
    public static StarnameAssets avax   = new StarnameAssets("Avalanche", "asset:avax", "AVAX", null, null);
    public static StarnameAssets band   = new StarnameAssets("Band Protocol", "asset:band", "BAND", "band.png", BaseChain.BAND_MAIN);
    public static StarnameAssets bat    = new StarnameAssets("Basic Attention Token", "asset:bat", "BAT", null, null);
    public static StarnameAssets bch    = new StarnameAssets("Bitcoin Cash", "asset:bch", "BCH", "bitcoincash.png", null);
    public static StarnameAssets bnb    = new StarnameAssets("BNB coin", "asset:bnb", "BNB", "binance.png", BaseChain.BNB_MAIN);
    public static StarnameAssets btc    = new StarnameAssets("Bitcoin", "asset:btc", "BTC", "bitcoin.png", null);
    public static StarnameAssets comp   = new StarnameAssets("Compound", "asset:comp", "COMP", null, null);
    public static StarnameAssets cro    = new StarnameAssets("Crypto.org", "asset:cro", "CRO", "crypto-org.png", BaseChain.CRYPTO_MAIN);
    public static StarnameAssets dai    = new StarnameAssets("Dai Stablecoin", "asset:dai", "DAI", null, null);
    public static StarnameAssets dash   = new StarnameAssets("Dash", "asset:dash", "DASH", null, null);
    public static StarnameAssets doge   = new StarnameAssets("Dogecoin", "asset:doge", "DOGE", null, null);
    public static StarnameAssets dsm    = new StarnameAssets("Desmos", "asset:dsm", "DSM", null, null);
    public static StarnameAssets dvpn   = new StarnameAssets("Sentinel", "asset:dvpn", "DVPN", "sentinel.png", BaseChain.SENTINEL_MAIN);
    public static StarnameAssets eos    = new StarnameAssets("EOS", "asset:eos", "EOS", null, null);
    public static StarnameAssets eth    = new StarnameAssets("Ethereum", "asset:eth", "ETH", "ethereum.png", null);
    public static StarnameAssets ion    = new StarnameAssets("ION", "asset:ion", "ION", "osmosis.png", BaseChain.OSMOSIS_MAIN);
    public static StarnameAssets iov    = new StarnameAssets("Starname (IOV)", "asset:iov", "IOV", "starname.png", BaseChain.IOV_MAIN);
    public static StarnameAssets iris   = new StarnameAssets("IRISnet", "asset:iris", "IRIS", "iris.png", BaseChain.IRIS_MAIN);
    public static StarnameAssets kava   = new StarnameAssets("Kava", "asset:kava", "KAVA", "kava.png", BaseChain.KAVA_MAIN);
    public static StarnameAssets ksm    = new StarnameAssets("Kusama", "asset:ksm", "KSM", null, null);
    public static StarnameAssets link   = new StarnameAssets("Chainlink", "asset:link", "LINK", null, null);
    public static StarnameAssets ltc    = new StarnameAssets("Litecoin", "asset:ltc", "LTC", "litecoin.png", null);
    public static StarnameAssets neo    = new StarnameAssets("Neo", "asset:neo", "NEO", null, null);
    public static StarnameAssets okb    = new StarnameAssets("OKB", "asset:okb", "OKB", "okex.png", BaseChain.OKEX_MAIN);
    public static StarnameAssets osmo   = new StarnameAssets("Osmosis", "asset:osmo", "OSMO", "osmosis.png", BaseChain.OSMOSIS_MAIN);
    public static StarnameAssets regen  = new StarnameAssets("Regen Network", "asset:regen", "REGEN", "regen.png", BaseChain.REGEN_MAIN);
    public static StarnameAssets ren    = new StarnameAssets("Ren", "asset:ren", "REN", null, null);
    public static StarnameAssets req    = new StarnameAssets("Request", "asset:req", "REQ", null, null);
    public static StarnameAssets rowan  = new StarnameAssets("Sifchain", "asset:rowan", "ROWAN", "sifchain.png", BaseChain.SIF_MAIN);
    public static StarnameAssets scrt   = new StarnameAssets("Secret", "asset:scrt", "SCRT", null, BaseChain.SECRET_MAIN);
    public static StarnameAssets shib   = new StarnameAssets("Shiba Inu", "asset:shib", "SHIB", null, null);
    public static StarnameAssets sol    = new StarnameAssets("Solana", "asset:sol", "SOL", null, null);
    public static StarnameAssets svt    = new StarnameAssets("Savitar Token", "asset:svt", "SVT", null, null);
    public static StarnameAssets trx    = new StarnameAssets("TRON", "asset:trx", "TRX", null, null);
    public static StarnameAssets usdc   = new StarnameAssets("USD Coin", "asset:usdc", "USDC", null, null);
    public static StarnameAssets usdt   = new StarnameAssets("Tether", "asset:usdt", "USDT", null, null);
    public static StarnameAssets wbtc   = new StarnameAssets("Wrapped Bitcoin", "asset:wbtc", "WBTC", null, null);
    public static StarnameAssets xlm    = new StarnameAssets("Stellar", "asset:xlm", "XLM", null, null);
    public static StarnameAssets xmr    = new StarnameAssets("Monero", "asset:xmr", "XMR", null, null);
    public static StarnameAssets xprt   = new StarnameAssets("Persistence", "asset:xprt", "XPRT", "persistence.png", BaseChain.PERSIS_MAIN);
    public static StarnameAssets xtz    = new StarnameAssets("Tezos", "asset:xtz", "XTZ", null, null);
    public static StarnameAssets zec    = new StarnameAssets("Zcash", "asset:zec", "ZEC", null, null);
}
